core/trino-main/src/main/java/io/trino/sql/planner/optimizations/DetermineHashPartitionCount.java

```diff
@@ -0,0 +1,255 @@
+import io.trino.cost.StatsProvider;
+import io.trino.cost.TableStatsProvider;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.sql.planner.plan.JoinNode;
+import io.trino.sql.planner.plan.MergeWriterNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+import io.trino.sql.planner.plan.TableExecuteNode;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.sql.planner.plan.TableWriterNode;
+import io.trino.sql.planner.plan.UnnestNode;
+import io.trino.sql.planner.plan.ValuesNode;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static io.trino.SystemSessionProperties.getMaxHashPartitionCount;
+import static io.trino.SystemSessionProperties.getMinInputRowsPerTask;
+import static io.trino.SystemSessionProperties.getQueryMaxMemoryPerNode;
+import static io.trino.SystemSessionProperties.getTaskPartitionedWriterCount;
+import static io.trino.cost.AggregationStatsRule.getRowsCount;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isAtMostScalar;
+import static io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static java.lang.Double.NaN;
+import static java.lang.Double.isNaN;
+import static java.lang.Math.ceil;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Estimate the value of hash partition count based on statistics.
+ */
+public class DetermineHashPartitionCount
+        implements PlanOptimizer
+{
+    private static final Logger log = Logger.get(DetermineHashPartitionCount.class);
+    private static final long WRITER_MEMORY_ALLOCATED_PER_PARTITION = DataSize.of(16, MEGABYTE).toBytes();
+    private static final int MAX_OPEN_PARTITIONS_PER_WRITER_THREAD = 100;
+
+    private final StatsCalculator statsCalculator;
+
+    public DetermineHashPartitionCount(StatsCalculator statsCalculator)
+    {
+        this.statsCalculator = requireNonNull(statsCalculator, "statsCalculator is null");
+    }
+
+    @Override
+    public PlanNode optimize(
+            PlanNode plan,
+            Session session,
+            TypeProvider types,
+            SymbolAllocator symbolAllocator,
+            PlanNodeIdAllocator idAllocator,
+            WarningCollector warningCollector,
+            TableStatsProvider tableStatsProvider)
+    {
+        requireNonNull(plan, "plan is null");
+        requireNonNull(session, "session is null");
+        requireNonNull(types, "types is null");
+        requireNonNull(tableStatsProvider, "tableStatsProvider is null");
+
+        try {
+            return determineHashPartitionCount(plan, session, types, tableStatsProvider)
```

Comment:
Would it be possible and beneficial to have different partition counts per different exchange nodes?

Reply:
We've tried that and come to the conclusion that it's not possible as of now and is hard to implement.